exporter/metric/metric.go

```diff
@@ -0,0 +1,384 @@
+// Copyright 2020, Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+
+	"go.opentelemetry.io/otel/api/global"
+	apimetric "go.opentelemetry.io/otel/api/metric"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+	"go.opentelemetry.io/otel/sdk/export/metric/aggregator"
+	"go.opentelemetry.io/otel/sdk/metric/controller/push"
+	integrator "go.opentelemetry.io/otel/sdk/metric/integrator/simple"
+	"go.opentelemetry.io/otel/sdk/metric/selector/simple"
+	"go.opentelemetry.io/otel/sdk/resource"
+
+	monitoring "cloud.google.com/go/monitoring/apiv3"
+	"google.golang.org/api/option"
+	googlemetricpb "google.golang.org/genproto/googleapis/api/metric"
+	monitoredrespb "google.golang.org/genproto/googleapis/api/monitoredres"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+)
+
+const (
+	version                                   = "0.1.0"
+	cloudMonitoringMetricDescriptorNameFormat = "custom.googleapis.com/opentelemetry/%s"
+)
+
+var (
+	errBlankProjectID         = errors.New("expecting a non-blank ProjectID")
+	errUnsupportedAggregation = errors.New("Currently the metric type is not supported")
+)
+
+// metricExporter is the implementation of OpenTelemetry metric exporter for
+// Google Cloud Monitoring.
+type metricExporter struct {
+	o *options
+
+	// mdCache is the cache to hold MetricDescriptor to avoid creating duplicate MD.
+	// TODO: [ymotongpoo] this map should be goroutine safe with mutex.
+	mdCache map[string]*googlemetricpb.MetricDescriptor
+
+	client *monitoring.MetricClient
+}
+
+// newMetricExporter returns an exporter that uploads OTel metric data to Google Cloud Monitoring.
+// Only one Cloud Monitoring exporter should be created per ProjectID per process, any subsequent
+// invocations of NewExporter with the same ProjectID will return an error.
```

Comment:
Does it already return an error in that case? I don't see any checks for existing exporters per ProjectID.

Reply:
I admit that I copy and pasted from OC's comment, and it's not handling such a case. For now, I eliminated the 2nd line and after.